src/source-map.ts

```diff
@@ -0,0 +1,192 @@
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as acorn from 'acorn';
+import * as dom5 from 'dom5';
+import * as parse5 from 'parse5';
+import {AnalysisContext} from 'polymer-analyzer/lib/core/analysis-context';
+import {RawSourceMap, SourceMapConsumer, SourceMapGenerator} from 'source-map';
+import * as urlLib from 'url';
+
+import * as matchers from './matchers';
+
+const inlineSourcemapPrefix =
+    '\n//# sourceMappingURL=data:application/json;charset=utf8;base64,';
+
+const sourceMappingUrlExpr = /\n\/\/# sourceMappingURL=(.*)\n?/;
+
+/**
+ * Creates an identity source map from JS script content. Can offset original
+ * line/column data for inline script elements.
+ */
+export function createJsIdentitySourcemap(
+    sourceUrl: string,
+    sourceContent: string,
+    lineOffset: number = 0,
+    firstLineCharOffset: number = 0) {
+  const generator = new SourceMapGenerator();
+  const tokenizer = acorn.tokenizer(sourceContent, {locations: true});
+  while (true) {
+    const token = tokenizer.getToken();
+    if (token.type.label === 'eof')
+      break;
+
+    if (!token.loc) {
+      return null;
+    }
+    let mapping: any = {
+      original: {
+        line: token.loc.start.line + lineOffset,
+        column: token.loc.start.column +
+            (token.loc.start.line === 1 ? firstLineCharOffset : 0)
+      },
+      generated: token.loc.start,
+      source: sourceUrl
+    };
+    if (token.type.label === 'name') {
+      mapping.name = token.value;
+    }
+
+    generator.addMapping(mapping);
+  }
+
+  return generator.toJSON();
+}
+
+const sourceMapCommentExpr =
+    /\n\/\/# sourceMappingURL=data:application\/json;(charset=[^;]+;)?base64,([a-zA-Z0-9+\/=]+)\n?\s*$/;
+
+function offsetSourceMap(
+    sourcemap: RawSourceMap, lineOffset: number, firstLineCharOffset: number) {
+  const consumer = new SourceMapConsumer(sourcemap);
+  const generator = new SourceMapGenerator();
+
+  consumer.eachMapping(mapping => {
+    const newMapping: any = {
+      source: mapping.source,
+      original: {line: mapping.originalLine, column: mapping.originalColumn},
+      generated: {
+        line: mapping.generatedLine + lineOffset,
+        column: mapping.generatedColumn +
+            (mapping.generatedLine === 1 ? firstLineCharOffset : 0)
+      }
+    };
+
+    if (mapping.name) {
+      newMapping.name = mapping.name;
+    }
+
+    generator.addMapping(newMapping);
+  });
+
+  return generator.toJSON();
+}
+
+export async function getExistingSourcemap(
+    analyzer: AnalysisContext, sourceUrl: string, sourceContent: string) {
+  const dataUrlPrefix = 'data:application/json;';
+  const sourceMappingUrlParts = sourceMappingUrlExpr.exec(sourceContent);
```

Comment:
I'm spoiled - Closure-compiler has long warned about any use of RegExp that depends on the global state. I don't even think about that any more ...